src/ffi/python/mod.rs

```diff
@@ -176,17 +200,127 @@ impl Config {
 			let port = kwgs.get_item("port")?.and_then(|e| e.extract().ok());
 			let tls = kwgs.get_item("tls")?.and_then(|e| e.extract().ok());
 
-			Ok(Config {
+			Ok(Self {
 				username,
 				password,
 				host,
 				port,
 				tls,
 			})
 		} else {
-			Ok(Config::new(username, password))
+			Ok(Self::new(username, password))
 		}
 	}
+
+	fn __str__(&self) -> String {
+		format!("{self:?}")
+	}
+}
+
+#[pymethods]
+impl Cursor {
+	fn __str__(&self) -> String {
+		format!("{self:?}")
+	}
+}
+
+#[pymethods]
+impl Selection {
+	#[new]
+	#[pyo3(signature = (**kwds))]
+	pub fn py_new(kwds: Option<&Bound<'_, PyDict>>) -> PyResult<Self> {
+		if let Some(kwds) = kwds {
+			let start_row = if let Some(e) = kwds.get_item("start_row")? {
+				e.extract()?
+			} else {
+				0
+			};
+
+			let start_col = if let Some(e) = kwds.get_item("start_col")? {
+				e.extract()?
+			} else {
+				0
+			};
+
+			let end_row = if let Some(e) = kwds.get_item("end_row")? {
+				e.extract()?
+			} else {
+				0
+			};
+
+			let end_col = if let Some(e) = kwds.get_item("end_col")? {
+				e.extract()?
+			} else {
+				0
+			};
+
+			let buffer = if let Some(e) = kwds.get_item("buffer")? {
+				e.extract()?
+			} else {
+				String::default()
+			};
+
+			Ok(Self {
+				start_row,
+				start_col,
+				end_row,
+				end_col,
+				buffer,
+			})
+		} else {
+			Ok(Self::default())
+		}
+	}
+
+	fn __str__(&self) -> String {
+		format!("{self:?}")
+	}
+}
+
+#[pymethods]
+impl BufferUpdate {
+	fn __str__(&self) -> String {
+		format!("{self:?}")
+	}
+}
+
+#[pymethods]
+impl TextChange {
+	#[new]
+	#[pyo3(signature = (**kwds))]
+	pub fn py_new(kwds: Option<&Bound<'_, PyDict>>) -> PyResult<Self> {
+		if let Some(kwds) = kwds {
+			let start_idx = if let Some(e) = kwds.get_item("start")? {
+				e.extract()?
+			} else {
+				0
+			};
+
+			let end_idx = if let Some(e) = kwds.get_item("end")? {
+				e.extract()?
+			} else {
+				0
+			};
+
+			let content = if let Some(e) = kwds.get_item("content")? {
+				e.extract()?
+			} else {
+				String::default()
+			};
+
+			Ok(Self {
+				start_idx,
+				end_idx,
+				content,
+			})
+		} else {
+			Ok(Self::default())
+		}
+	}
+
+	fn __str__(&self) -> String {
+		format!("{self:?}")
+	}
```

Comment:
Why all these manual constructors?
Not giving the user direct ability to construct such internal objects allowed for input validations at the plugin level.

Reply:
What do you mean by "manual constructor"? `__str__`'s purpose is to allow doing str(object) to see its internal makeup.

Reply:
assuming the mentioned code is wrong and we're referring to the #[new] methods we added, we did it this way for mostly three reasons:
 * consistency: all other languages have `.send()` methods receive a single object, the thing they're sending, rather than arguments to build it
 * comfort: a delete-only change can avoid setting its content. "out-of-buffer" cursors can just send a `codemp.Selection()` and send defaults, rather than putting manually `0, 0, ""`
 * maintainability: if we change the struct, we don't need to update method signatures. right now, we would still need to update the constructor, but it's a single place rather than many (well actually it's passed only to `.send()`, but i wouldn't count on it remaining that way forever). Also pyo3 has a `serde` feature which I think we can leverage to avoid doing constructors. in lua, everything that is serializable gets automagically translated from/into a lua table, so we just change the struct definition and it impacts everything else without extra work. js does something similar with `napi(object)` and i dream of doing something equivalent in `jni-toolbox`. making constructors allows us to expose the "future" api right away and eventually figure this out later with just a patch version, and worst case we just moved the validation logic inside constructor rather than `.send()` itself, while other two points still hold
 
Hope this was what you meant!

Reply:
Yes I meant to address that whole hunk of code with the comment, non just the str, sorry.
Ok I see the point, and now that I think of it, the point I raised about input validation is wrong. This way we have much more control since it is done in the glue or library instead of in the plugin.

Sold.